Fix locally triggered purge

milan-zededa opened the pull request:
This PR fixes the feature allowing to trigger app purge/restart via 
local profile server, originally delivered by https://github.com/lf-edge/eve/pull/2492
As it turned out, to purge an application _with all its volumes_,
it is required to either change UUIDs of those volumes (effectively
configure new volume instances) or to increase their generation counters.
However, locally triggered application commands (via local profile server)
cannot change configuration in a way that would break consistency with
the controller, therefore change of UUIDs is not an option.
Instead, this PR introduces a _local generation counter_ for volume,
which is added to the remote generation counter (from the controller)
to form a volume key changing by remotely as well as locally issued purge.
In a similar way the PR adds separate purge and restart counters
for locally triggered operations to the application config.
Note that currently most of the logic for local operations is handled
by zedagent. Later, this could be refactored and moved to zedmanager
to keep only config parsing and info/metrics publishing inside zedagent.

Test: https://github.com/lf-edge/eden/pull/744

Signed-off-by: Milan Lenco <milan@zededa.com>

giggsoff commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -174,7 +173,11 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 	var cmdChanges bool
+	publishApps := make(map[string]types.AppInstanceConfig)
+	publishVols := make(map[string]types.VolumeConfig)
+	unpublishVols := make(map[string]struct{})
```
Seems slice is enough.

milan-zededa replied:
Replaced maps with slices as suggested.

giggsoff commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,73 +195,133 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
+					continue
+				}
+				unpublishVols[volKey] = struct{}{}
+				vr.GenerationCounter++
+				volume := volObj.(types.VolumeConfig)
+				volume.GenerationCounter++
+				publishVols[volume.Key()] = volume
+				appCmd.VolumeGenCounter[vr.VolumeID.String()]++
```
Should we have configurable option in api to determine which volumes we should keep not-purgeable during this operation?

eriknordmark replied:
I think we need to add this; the commercial controller already has such a notion internally.
But we need to check what semantics the users want for the local purge - whether they want to purge all the volumes or only the ones which the controller has marked as "purgeable" in some API addition.

eriknordmark replied:
How about adding documentation to the API document to state that the purge command will purge all the volumes. That way we don't forget.

milan-zededa replied:
Added comment about this to `local_profile.proto`.

eriknordmark commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,73 +195,133 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
+		switch appCmd.Command {
+		case types.AppCommandRestart:
+			appCmd.RestartCounter++
+			appInst.RestartCmd.Counter++
+			publishApps[appUUID.String()] = *appInst
+
+		case types.AppCommandPurge:
+			appCmd.PurgeCounter++
+			appInst.PurgeCmd.Counter++
```
When we get a restart or purge command (well, actally an update to the AppInstanceConfig) from the controller we will check against the previous counters from the controller. Won't this local restart/purge command upset that check since they appear to increment the counters (as opposed to a separate Local*Counter?)

eriknordmark replied:
Yeah this is very confusing.

IMHO zedmanager should operate on the AppInstanceConfig received from the controller.
Thus zedmanager would need to add in the local cmd counters as appropriate.

The original code (where zedagent looks at AppInstanceStatus to report back to the LPS) is fine, but purging/restarting is sufficiently complex to not have that logic spread between zedagent and zedmanager.

eriknordmark commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -290,10 +351,24 @@ func prepareLocalInfo(ctx *getconfigContext) *profile.LocalAppInfoList {
+// localAppCommandsLock should be locked when calling this function!
+func getLocalGenCountForVolume(ctx *getconfigContext, volumeID string) int64 {
+	for _, localCmd := range ctx.localAppCommands {
+		if counter, match := localCmd.VolumeGenCounter[volumeID]; match {
```
Is this localCmd persisted across reboots of the device?

eriknordmark replied:
If so a comment about that would be helpful

milan-zededa replied:
Yes it is persisted, I will add a comment.

milan-zededa replied:
done

rvs reviewed (approved):
LGTM! Great solution @milan-zededa

eriknordmark commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,73 +197,135 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 		// Try to find the application instance.
-		ais := findAppInstance(ctx, appUUID, displayName)
-		if ais == nil {
+		appInst := findAppInstance(ctx, appUUID, displayName)
```
This is the AppInstanceConfig received from zedagent, right?
It is confusing to modify that (the appInst.RestartCmd.Counter++ etc below) since now you are locally having something which is different than what was published (and it already confused me).

eriknordmark replied:
Ah - this is in zedagent. Now I'm even more confused. zedagent is normally just protobuf to json conversion and here you have purge/restart logic. Can't that logic live in zedmanager?

milan-zededa replied:
The only problem is the VolumeConfig, which goes directly from zedagent to volummgr. I would need to "re-route" it to go through zedmanger (to get local gen. counter). Do you see any potential problem this could introduce?

eriknordmark replied:
We some factoring out of the current code I think having this in zedagent is fine for now.

milan-zededa commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,73 +197,135 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 		touchAppCommands()
 	}
+
+	// Publish updated configuration.
+	for _, volKey := range unpublishVols {
+		unpublishVolumeConfig(ctx, volKey)
+	}
+	for _, volume := range publishVols {
+		publishVolumeConfig(ctx, volume)
+	}
```
OK, I will split the counters then.

milan-zededa commented:
@eriknordmark I have refactored the PR and split local counters into separate fields to make it less confusing. I have also done some renaming and changes to make this ready for next commands that we plan to support in offline mode.
Still the merging of the local and the remote config is done in zedagent. This is primarily because it is much easier to get it right. Moving this to zedmanager is susceptible to race conditions (local and remote config would be published separately and timing might cause problems). Later, as we work on offline services for our next big release, we might move this logic to zedmanager, but currently I do not feel confident doing this since I do not have a detail knowledge of volumemgr and how it interacts with zedmanager and other microservices. Unfortunately I do not have the capacity to familiarize myself with this in a timely manner and there is a demand to get the local-purge functionality into EVE ASAP (to master and also backport to 7.9).

eriknordmark commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -58,17 +61,14 @@ func (config VolumeConfig) LogModify(logBase *base.LogObject, old interface{}) {
 	if oldConfig.ContentID != config.ContentID ||
 		oldConfig.MaxVolSize != config.MaxVolSize ||
-		oldConfig.RefCount != config.RefCount ||
-		oldConfig.GenerationCounter != config.GenerationCounter {
+		oldConfig.RefCount != config.RefCount {
 
 		logObject.CloneAndAddField("content-id", config.ContentID).
 			AddField("max-vol-size-int64", config.MaxVolSize).
 			AddField("refcount-int64", config.RefCount).
-			AddField("generation-counter-int64", config.GenerationCounter).
 			AddField("old-content-id", oldConfig.ContentID).
 			AddField("old-max-vol-size-int64", oldConfig.MaxVolSize).
 			AddField("old-refcount-int64", oldConfig.RefCount).
-			AddField("old-generation-counter-int64", oldConfig.GenerationCounter).
```
Don't you want to log changes to the local and normal generation counter?

milan-zededa replied:
The counters are part of the key so any change in them will trigger Delete/Create, not Modify.

eriknordmark commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,81 +203,199 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
+				volume := volObj.(types.VolumeConfig)
+				unpublishVols = append(unpublishVols, volKey)
+				// Publish volume with an increased local generation counter.
+				localGenCounter++
```
If an app instance has multiple volumes you increase this for each one. That seems unnecessary. 
Shouldn't you instread just increment (and set to 1 if doesn't exist) ctx.localCommands.VolumeGenCounters[uuid]? Or pick a localGenCounter value for the app instance and add that for all of the volumes?

eriknordmark replied:
OK

eriknordmark commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,81 +203,199 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
+		case types.AppCommandPurge:
+			appCounters.PurgeCmd.Counter++
+			appCounters.PurgeCmd.ApplyTime = appCmd.DeviceTimestamp.String()
+			appInst.LocalPurgeCmd = appCounters.PurgeCmd
```
I would add a comment here (or maybe a comment block above) that we Get the AppInstanceConfig and update it wih a new generation counter (and we also do that for the VolumeConfig/VolumeRefConfig for the app instance) and then re-publish those.

milan-zededa replied:
Done - added separate explanation for Restart and for Purge.

eriknordmark commented on pkg/pillar/cmd/zedagent/localinfo.go:
```diff
@@ -192,81 +203,199 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
+		appCmd.DeviceTimestamp = time.Now()
+		appCmd.Completed = false
+
+		appCounters, hasCounters := ctx.localCommands.AppCounters[appUUID.String()]
```
It feels like we should be able to have an abstration/function here which takes care of updating the appInst+volumeref+volume and republish. Presumably it wouldn't need anything but appCmd.Command and it wouldn't modify appCmd?

milan-zededa replied:
Done - added a new function `triggerLocalCommand`
